### PR TITLE
#1473: Feature Request: Admin Side Tagging

### PR DESCRIPTION
**Pull request by aschneit**
# Fixes # (issue)
#1473 
## Description

Implemented UI for Tags admin screen and related CRUD functionality (creating, editing, deleting tags). 

# Checklist:

- [x] I have manually tested my changes on desktop and mobile
- [x] The test suite passes locally with my changes
- [x] If my change is a UI change, I have attached a screenshot to the description section of this pull request
- [x] [My change is 300 lines of code or less](https://github.com/MoveOnOrg/Spoke/blob/main/CONTRIBUTING.md#your-first-code-contribution), or has a documented reason in the description why it’s longer
- [x] I have made any necessary changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] My PR is labeled [WIP] if it is in progress

Screen Shot 2020-04-23 at 11 54 27 AM

### CODE REVIEW

**Comment by azuzunaga** [comment]
Taking a look now!

File: src/containers/Tags.jsx
Comment (azuzunaga) [inline src/containers/Tags.jsx]: Would it make more sense to have `handleOpen` and `handleOpenEdit` set state values for `title`, `submitHandler`, and `buttonLabel`? It seems to me that would be cleaner (and more extensible) than the ternary expressions.

File: src/server/api/schema.js
Comment (azuzunaga) [inline src/server/api/schema.js:1254]: Maybe I missed something in the issue conversation, but why don't we actually delete the tags? Is it to avoid issues with cascading deletes in the `tag_campaign_contact` table?

File: src/server/models/index.js
Comment (azuzunaga) [inline src/server/models/index.js]: Nit:
```suggestion
  "zip_code" // good candidate (or by contact)?
```

**Review by azuzunaga — COMMENTED** [review]
Great work on this! This feature has been requested a bunch, glad it is getting implemented!

I have a few questions before I approve, let me know what you think.

Thanks!

File: src/server/middleware/render-index.js
Comment (aschneit) [reply src/server/middleware/render-index.js]: Seems like the other boolean vars above (NOT_IN_USA and ALLOW_SEND_ALL) have 0 as their default if the variable isn't set. But I can change it if that doesn't seem right. (also, this environment variable is also just temporary, so that Tags option won't show up in the admin sidebar until we get the texter tag ui in place)

File: src/components/AdminDashboard.jsx
Comment (aschneit) [reply src/components/AdminDashboard.jsx]: See above, since this is probably just a temporary variable, I thought just being able to remove the false conditional when we no longer need it would be easier. But I'm open to changing it if it seems messy. @ibrand, we'll probably just remove this env. variable when we get the texter tag ui set to go,  right?

File: src/server/api/schema.js
Comment (aschneit) [reply src/server/api/schema.js:1254]: @ibrand  gave me part of the explanation for this, but I wasn't totally clear on it and there may be more that I'm not aware of. Ilona?

File: src/api/schema.js
Comment (aschneit) [reply src/api/schema.js]: removed that

File: src/containers/Tags.jsx
Comment (aschneit) [reply src/containers/Tags.jsx]: added those, thanks

File: src/server/models/index.js
Comment (aschneit) [reply src/server/models/index.js]: removed

File: src/server/middleware/render-index.js
Comment (azuzunaga) [reply src/server/middleware/render-index.js]: Gotcha. Let's set it to `false` then, since we check explicitly for a `false` value in the admin dashboard.

File: src/components/AdminDashboard.jsx
Comment (azuzunaga) [reply src/components/AdminDashboard.jsx]: If we are removing that if block anyway then we can leave it as is. Thanks!

**Review by azuzunaga — DISMISSED** [review]
Looks great @aschneit! Thanks for pushing up the changes so quickly. I just left a final comment on the `window.SHOW_TAGS=${process.env.SHOW_TAGS || 0}` piece of code, but it is not a blocker.

I'm interested to hear about the `is_deleted` field, but again, not blocking.

🥳

**Review by azuzunaga — DISMISSED** [review]
Looks like the last commit dismissed the previous approval 🙂

File: src/server/api/schema.js
Comment (ibrand) [reply src/server/api/schema.js:1254]: Hi @aschneit and @azuzunaga thanks for the ping :)

I encouraged a soft delete because if tags are soft deleted but still exist attached to other things (like canned responses) then no other data will be messed up. Secondarily, I figure an admin could accidentally delete one and want it back.

I think that first situation is a lot more important because it could lead to some bad data situations for a tag to no longer exist but a lot of the data export has tags already in it.

Comment (azuzunaga) [reply src/server/api/schema.js:1254]: Thanks for the explanation @ibrand! Makes sense to me.

File: src/server/models/tag-campaign-contact.js
Comment (matteosb) [inline src/server/models/tag-campaign-contact.js:26]: These don't do anything right now and knex migrations are the source of truth, so I would remove them 

cc/ @schuyler1d. If we want to continue adding these for some reason, then let's merge this PR: https://github.com/MoveOnOrg/rethink-knex-adapter/pull/28

**Review by schuyler1d — DISMISSED** [review]
:+1:

**Comment by aschneit** [comment]
@schuyler1d, @azuzunaga, @matteosb, I believe I resolved all of your comments (but the approvals seem to have disappeared after I pushed new commits). The only one I didn't address is the one about ensureIndex in the TagCampaignContact model, since that still seemed like an open question contingent on merging that other pr...and the models/migrations weren't commits I actually worked on. But I can make the change if people want, just let me know, thanks!

File: src/containers/Tags.jsx
Comment (aschneit) [inline src/containers/Tags.jsx:106]: @matteosb I added this line in all of the mutation event handlers to refetch data after the mutation. Without it, dom updates don't seem to work. But if I made the changes you suggested above for updating the cache correctly after a mutation, should a refetch be necessary? My experience with GraphQL is limited, so I may be missing something. Thanks

Comment (matteosb) [reply src/containers/Tags.jsx:106]: Interesting, this shouldn't be necessary if you are returning the fields you are updating from the mutation but we are currently on an ancient version of apollo-client, so it might be broken. It's fine to leave this in here for now. I'm going to do a pass on this once I upgrade apollo-client (see the discussion here: #1461)

Comment (aschneit) [reply src/containers/Tags.jsx:106]: @matteosb Yeah, I noticed there isn't an apollo-cache package installed, and I wasn't totally clear on the caching strategies being used overall. I guess we'll see what can be changed with the upgrade. Thanks!

**Comment by schuyler1d** [comment]
I believe everything in this PR was merged into what will be released as Spoke 5.5.  I think commit order meant that it isn't showing as merged, but it is.

**Pull request closed without merging**
